operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java

```diff
@@ -275,95 +337,136 @@ public CertAndKey generateSignedCert(String commonName, String organization) thr
+        if (caCertsRemoved) {
             log.info("{}: Expired CA certificates removed", this);
         }
-        if (caRenewed) {
-            log.info("{}: Certificates renewed", this);
+        if (renewalType != RenewalType.NOOP) {
+            log.debug("{}: {}", this, renewalType.postDescription(caKeySecretName, caCertSecretName));
         }
 
         // cluster CA certificate generation annotation handling
-        int caCertGeneration = INIT_GENERATION;
-        if (caCertSecret != null && caCertSecret.getData().get(CA_CRT) != null) {
-            caCertGeneration = Annotations.intAnnotation(caCertSecret, ANNO_STRIMZI_IO_CA_CERT_GENERATION,
-                    INIT_GENERATION);
-            if (caRenewed) {
-                caCertGeneration++;
-            }
-        }
-
         caCertSecret = secretCertProvider.createSecret(namespace, caCertSecretName, certData, labels,
                 Collections.singletonMap(ANNO_STRIMZI_IO_CA_CERT_GENERATION, String.valueOf(caCertGeneration)), ownerRef);
+
         caKeySecret = secretCertProvider.createSecret(namespace, caKeySecretName, keyData, labels,
-                Collections.emptyMap(), ownerRef);
+                Collections.singletonMap(ANNO_STRIMZI_IO_CA_KEY_GENERATION, String.valueOf(caKeyGeneration)), ownerRef);
     }
 
-    private boolean shouldCreateOrRenew(X509Certificate currentCert, String namespace, String clusterName) {
+    private Subject nextCaSubject(int version) {
+        Subject result = new Subject();
+        // Key replacements does not work if both old and new CA certs have the same subject DN, so include the
+        // key generation in the DN so the certificates appear distinct during CA key replacement.
+        result.setCommonName(commonName + " v" + version);
+        result.setOrganizationName(IO_STRIMZI);
+        return result;
+    }
+
+    private RenewalType shouldCreateOrRenew(X509Certificate currentCert, String namespace, String clusterName) {
         String reason = null;
-        boolean result = false;
-        if (this.caCertSecret == null
-                || this.caCertSecret.getData().get(CA_CRT) == null) {
-            reason = "CA certificate secret " + caCertSecretName + " is missing or lacking the key " + CA_CRT;
-            result = true;
-            this.caRenewed = this.caCertSecret != null;
-        } else if (caKeySecret == null
+        RenewalType renewalType = RenewalType.NOOP;
+        if (caKeySecret == null
                 || caKeySecret.getData().get(CA_KEY) == null) {
             reason = "CA key secret " + caKeySecretName + " is missing or lacking the key " + CA_KEY;
-            result = true;
-            this.caRenewed = caKeySecret != null;
+            renewalType = RenewalType.CREATE;
+        } else if (this.caCertSecret == null
+                || this.caCertSecret.getData().get(CA_CRT) == null) {
+            reason = "CA certificate secret " + caCertSecretName + " is missing or lacking the key " + CA_CRT;
```

Comment:
So there's an annoying overload in terminology here: The mappings in the `data` of a `Secret` have a 'key' (and a 'value'), but in this case we're talking about `Secrets` containing certificates and keys. I guess we can say "...is lacking spec.data.ca\.crt" to avoid the ambiguity.

Reply:
Whether to escape the `.` in `ca.crt` though? I suppose we should.

Reply:
No I think we are fine as it is.
It seems that the first time my eyes stopped reading at "...key" not considering the `CA_CRT` concatenation. Now it's clear to me that you are referring to the key `CA_CRT` as the "field name" of the `data` section of the `Secret`